File: native-android/app/src/main/java/site/smap/gugudan/designsystem/Feedback.kt
package site.smap.gugudan.designsystem

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import site.smap.gugudan.core.Achievements
import site.smap.gugudan.core.Level
import site.smap.gugudan.core.adventure.AdvAchievements
import kotlin.random.Random

// 피드백 오버레이 — Confetti / LevelUpOverlay / AchievementToast (iOS Feedback.swift 이식)

// 업적 아이콘 키 → Material 아이콘
fun achievementIcon(key: String): ImageVector = when (key) {
    "Check" -> Icons.Filled.Check
    "Target" -> Icons.Filled.TrackChanges
    "Award" -> Icons.Filled.EmojiEvents
    "Zap" -> Icons.Filled.Bolt
    "Flame" -> Icons.Filled.LocalFireDepartment
    "CalendarCheck", "CalendarHeart" -> Icons.Filled.Event
    "Crown" -> Icons.Filled.EmojiEvents
    "Star" -> Icons.Filled.Star
    "GraduationCap" -> Icons.Filled.School
    "Sparkles" -> Icons.Filled.AutoAwesome
    "Trophy" -> Icons.Filled.EmojiEvents
    "TrendingUp" -> Icons.Filled.TrendingUp
    "Rocket" -> Icons.Filled.RocketLaunch
    "Timer" -> Icons.Filled.Timer
    "HeartPulse" -> Icons.Filled.Favorite
    "Compass" -> Icons.Filled.Explore
    "Swords" -> Icons.Filled.Shield
    "Gauge" -> Icons.Filled.Speed
    "Medal" -> Icons.Filled.MilitaryTech
    "Flag" -> Icons.Filled.Flag
    "Gem" -> Icons.Filled.Diamond
    else -> Icons.Filled.Star
}

/** 업적 id → (이름, 아이콘) — 코어/어드벤처 통합 조회 */
fun achievementInfo(id: String): Pair<String, ImageVector>? {
    Achievements.get(id)?.let { return it.name to achievementIcon(it.icon) }
    AdvAchievements.get(id)?.let { return it.name to achievementIcon(it.icon) }
    return null
}

// MARK: Confetti

private data class ConfettiPiece(
    val x: Float, val color: Color, val delay: Float, val rotation: Float, val size: Float)

@Composable
fun ConfettiView(trigger: Int, modifier: Modifier = Modifier) {
    val colors = listOf(GGColors.indigo, GGColors.amber, GGColors.emerald, GGColors.rose, GGColors.sky, GGColors.violet)
    var pieces by remember { mutableStateOf<List<ConfettiPiece>>(emptyList()) }
    val progress = remember { Animatable(0f) }

    LaunchedEffect(trigger) {
        if (trigger <= 0) return@LaunchedEffect
        pieces = List(40) {
            ConfettiPiece(
                x = Random.nextFloat(), color = colors.random(),
                delay = Random.nextFloat() * 0.3f,
                rotation = Random.nextFloat() * 360f,
                size = 6f + Random.nextFloat() * 5f)
        }
        progress.snapTo(0f)
        progress.animateTo(1f, tween(1800))
        pieces = emptyList()
    }

    if (pieces.isNotEmpty()) {
        Canvas(modifier.fillMaxSize()) {
            val h = size.height
            for (p in pieces) {
                val t = ((progress.value - p.delay) / (1f - p.delay)).coerceIn(0f, 1f)
                if (t <= 0f) continue
                val y = -40f + (h + 80f) * t
                val alpha = 1f - t
                rotate(p.rotation + t * 360f, pivot = Offset(p.x * size.width, y)) {
                    drawRect(
                        p.color.copy(alpha = alpha),
                        topLeft = Offset(p.x * size.width - p.size / 2, y - p.size),
                        size = Size(p.size.dp.toPx() / 2.5f, p.size.dp.toPx() / 1.6f),
                    )
                }
            }
        }
    }
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.rotate(
    degrees: Float, pivot: Offset, block: androidx.compose.ui.graphics.drawscope.DrawScope.() -> Unit,
) = drawContext.transform.let { tf ->
    tf.rotate(degrees, pivot); block(); tf.rotate(-degrees, pivot)
}

// MARK: LevelUpOverlay

@Composable
fun LevelUpOverlay(show: Boolean, level: Int, onClose: () -> Unit) {
    if (!show) return
    val gg = LocalGG.current
    LaunchedEffect(Unit) { site.smap.gugudan.services.Haptics.success() }
    Box(
        Modifier.fillMaxSize().background(Color.Black.copy(alpha = 0.6f))
            .clickable(onClick = onClose),
        contentAlignment = Alignment.Center,
    ) {
        Column(
            Modifier.clip(RoundedCornerShape(28.dp)).background(gg.surface).padding(40.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Icon(Icons.Filled.AutoAwesome, null, tint = gg.warning, modifier = Modifier.size(44.dp))
            Text("레벨 업!", style = suite(FontWeight.ExtraBold, 28), color = gg.text)
            Text("Lv.$level", style = suite(FontWeight.Black, 48), color = gg.accent)
            Text(Level.title(level), style = suite(FontWeight.Bold, 15), color = gg.textMuted)
        }
    }
}

// MARK: AchievementToast

@Composable
fun AchievementToast(ids: List<String>, modifier: Modifier = Modifier) {
    var visible by remember(ids) { mutableStateOf(ids) }
    val gg = LocalGG.current

    LaunchedEffect(ids) {
        if (ids.isEmpty()) return@LaunchedEffect
        site.smap.gugudan.services.Haptics.success()
        delay(2800)
        visible = emptyList()
    }

    Column(modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        visible.forEach { id ->
            achievementInfo(id)?.let { (name, icon) ->
                Row(
                    Modifier.fillMaxWidth().ggCard(16.dp).padding(12.dp),
                    horizontalArrangement = Arrangement.spacedBy(10.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Box(Modifier.size(40.dp).clip(RoundedCornerShape(12.dp))
                        .background(gg.warning.copy(alpha = 0.15f)), contentAlignment = Alignment.Center) {
                        Icon(icon, null, tint = gg.warning, modifier = Modifier.size(20.dp))
                    }
                    Column {
                        Text("업적 달성", style = suite(FontWeight.Bold, 11), color = gg.textMuted)
                        Text(name, style = suite(FontWeight.ExtraBold, 15), color = gg.text)
                    }
                }
            }
        }
    }
}

File: native-ios/Gugu/DesignSystem/Feedback.swift
import SwiftUI

// 피드백 오버레이 — Confetti / LevelUpOverlay / AchievementToast (components/feedback 이식)

// MARK: - lucide 아이콘명 → SF Symbols 매핑 (업적 아이콘)

enum IconMap {
    static func sf(_ lucide: String) -> String {
        switch lucide {
        case "Check": return "checkmark"
        case "Target": return "target"
        case "Award": return "rosette"
        case "Zap": return "bolt.fill"
        case "Flame": return "flame.fill"
        case "CalendarCheck": return "calendar.badge.checkmark"
        case "CalendarHeart": return "calendar"
        case "Crown": return "crown.fill"
        case "Star": return "star.fill"
        case "GraduationCap": return "graduationcap.fill"
        case "Sparkles": return "sparkles"
        case "Trophy": return "trophy.fill"
        case "TrendingUp": return "chart.line.uptrend.xyaxis"
        case "Rocket": return "airplane"
        case "Timer": return "timer"
        case "HeartPulse": return "heart.fill"
        case "Compass": return "safari.fill"
        case "Swords": return "shield.lefthalf.filled"
        case "Gauge": return "gauge.medium"
        case "Medal": return "medal.fill"
        case "Flag": return "flag.fill"
        case "Gem": return "diamond.fill"
        default: return "star.fill"
        }
    }

    /// 업적 id → (이름, SF 아이콘) — 코어/어드벤처 업적 통합 조회
    static func achievement(_ id: String) -> (name: String, icon: String)? {
        if let a = Achievements.get(id) { return (a.name, sf(a.icon)) }
        if let a = AdvAchievements.get(id) { return (a.name, sf(a.icon)) }
        return nil
    }
}

// MARK: - Confetti

struct ConfettiView: View {
    var trigger: Int
    private let colors: [Color] = [.gg.accent, .gg.warning, .gg.success, .gg.danger, .gg.sky, .gg.violet]

    struct Piece: Identifiable {
        let id = UUID()
        let x: CGFloat
        let color: Color
        let delay: Double
        let rotation: Double
        let size: CGFloat
    }
    @State private var pieces: [Piece] = []
    @State private var animate = false

    var body: some View {
        GeometryReader { geo in
            ZStack {
                ForEach(pieces) { p in
                    Rectangle()
                        .fill(p.color)
                        .frame(width: p.size, height: p.size * 1.6)
                        .rotationEffect(.degrees(animate ? p.rotation + 360 : p.rotation))
                        .position(x: p.x * geo.size.width,
                                  y: animate ? geo.size.height + 40 : -40)
                        .opacity(animate ? 0 : 1)
                        .animation(.easeIn(duration: 1.8).delay(p.delay), value: animate)
                }
            }
        }
        .allowsHitTesting(false)
        .onChange(of: trigger) { _, newValue in
            guard newValue > 0 else { return }
            spawn()
        }
    }

    private func spawn() {
        pieces = (0..<40).map { _ in
            Piece(x: CGFloat.random(in: 0...1), color: colors.randomElement()!,
                  delay: Double.random(in: 0...0.4), rotation: Double.random(in: 0...360),
                  size: CGFloat.random(in: 6...11))
        }
        animate = false
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.02) {
            animate = true
        }
    }
}

// MARK: - LevelUpOverlay

struct LevelUpOverlay: View {
    var show: Bool
    var level: Int
    var onClose: () -> Void

    var body: some View {
        if show {
            ZStack {
                Color.black.opacity(0.6).ignoresSafeArea().onTapGesture(perform: onClose)
                VStack(spacing: 12) {
                    Image(systemName: "sparkles").font(.system(size: 44)).foregroundStyle(Color.gg.warning)
                    Text("레벨 업!").font(.suite(.extrabold, 28)).foregroundStyle(.white)
                    Text("Lv.\(level)").font(.suite(.heavy, 48)).foregroundStyle(Color.gg.accent).monospacedDigit()
                    Text(Level.title(level)).font(.suite(.bold, 15)).foregroundStyle(.white.opacity(0.8))
                }
                .padding(40)
                .background(Color.gg.surface, in: RoundedRectangle(cornerRadius: 28, style: .continuous))
                .transition(.scale.combined(with: .opacity))
            }
            .onAppear { Haptics.success() }
        }
    }
}

// MARK: - AchievementToast

struct AchievementToast: View {
    var ids: [String]
    @State private var visible: [String] = []

    var body: some View {
        VStack(spacing: 8) {
            ForEach(visible, id: \.self) { id in
                if let a = IconMap.achievement(id) {
                    HStack(spacing: 10) {
                        Image(systemName: a.icon).font(.system(size: 18, weight: .bold))
                            .foregroundStyle(Color.gg.warning)
                            .frame(width: 40, height: 40)
                            .background(Color.gg.warning.opacity(0.15), in: RoundedRectangle(cornerRadius: 12, style: .continuous))
                        VStack(alignment: .leading, spacing: 1) {
                            Text("업적 달성").font(.suite(.bold, 11)).foregroundStyle(Color.gg.textMuted)
                            Text(a.name).font(.suite(.extrabold, 15)).foregroundStyle(Color.gg.text)
                        }
                        Spacer()
                    }
                    .padding(12)
                    .background(Color.gg.surface, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
                    .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous).strokeBorder(Color.gg.border, lineWidth: 1))
                    .shadow(color: .black.opacity(0.15), radius: 8, y: 4)
                    .transition(.move(edge: .top).combined(with: .opacity))
                }
            }
        }
        .onChange(of: ids) { _, _ in showAll() }
        .onAppear { showAll() }
    }

    private func showAll() {
        guard !ids.isEmpty else { return }
        withAnimation(.spring(response: 0.4, dampingFraction: 0.7)) { visible = ids }
        Haptics.success()
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.8) {
            withAnimation { visible = [] }
        }
    }
}
